### ISSUE TITLE
Pool example from the pg readme does not compile

### ISSUE THREAD

**Issue by 0815fox** [goenning/typed-pg#7]
[Readme.md of pg](https://www.npmjs.com/package/pg) shows following example to create a pool:

```
var config = {
  user: 'foo', //env var: PGUSER 
  database: 'my_db', //env var: PGDATABASE 
  password: 'secret', //env var: PGPASSWORD 
  port: 5432, //env var: PGPORT 
  max: 10, // max number of clients in the pool 
  idleTimeoutMillis: 30000, // how long a client is allowed to remain idle before being closed 
};


//this initializes a connection pool 
//it will keep idle connections open for a 30 seconds 
//and set a limit of maximum 10 idle clients 
var pool = new pg.Pool(config);
...
```

However, new pg.Pool() does not exist in the module.

My first try to fix it was, to simply re-export Pool inside of `~pg/index`. However, I could not yet figure out, what to pass in for the type T and the factory argument of the constructor. Due to the example, a simple config object shall be passed instead of a factory.

Any suggestions?

**Comment by 0815fox** [comment]
Could it be, that the type definition is for version 4.5.x?

**Comment by goenning** [comment]
It looks like that new version has changed how they work with pools. They now have an exclusive package for that: https://github.com/brianc/node-pg-pool

I'll try to update it later, but I'd happily accept a PR on this. 
We'll need to create a new type definition for node-pg-pool for export it from typed-pg.

**Comment by 0815fox** [comment]
You are right, they separated it to node-pg-pool. Shouldn't it then become an extra package on TSD as well? I'll try to fix it on my fork of TSD.

**Comment by 0815fox** [comment]
Why do you declare the modules with those `~ aliases` and then export them?

```
declare module 'pg/index' {
import alias = require('~pg/index');
export = alias;
}
declare module 'pg' {
import alias = require('~pg/index');
export = alias;
}
```

**Comment by goenning** [comment]
I don't. Take a look at the source code of [index.d.ts](https://github.com/goenning/typed-pg/blob/master/index.d.ts).

That module declaration is what [typings](https://github.com/typings/typings) does when it's installing a new typing definition.

**Comment by 0815fox** [comment]
Ah gotcha, stupid mistake. Thank you for clarification. I am pretty new to this.

**Comment by 0815fox** [comment]
The definition for pg-pool would be sth. like this:

```
import { Promise } from "es6-promise"

...
export class PgPool extends EventEmitter {
  connect(cb: ConnectCallback): Promise<Client>;
  take(cb: ConnectCallback): Promise<Client>;
  query(query: QueryConfig, callback?: QueryCallback): Promise<Query>;
  query(text: string, callback: QueryCallback): Promise<Query>;
  query(text: string, values: any[], callback: QueryCallback): Promise<Query>;
  end(cb: Done): Promise<void>;

  on(event: "connect", listener: (client: Client) => void): this;
  on(event: "acquire", listener: (client: Client) => void): this;
  on(event: "error", listener: (err: Error) => void): this;
  on(event: string, listener: Function): this;
}

export const Pool: new (config: Config) => PgPool;
```

Again the question: Would it make more sense to have it as a separate definition, as pg-pool is a separate module now? If so, do you want to open another project on gitlab for that?

**Comment by goenning** [comment]
I think it'd be better to have another typing project just for `pg-pool` (probably `typed-pg-pool`). Would you like to host it under your github account? I suppose you mistyped github with gitlab.

On `typed-pg` we would end up adding `typed-pg-pool` as a dependency and export the Pool class.

**Comment by 0815fox** [comment]
Sure, I meant github, sorry.
I could do that. How is such a project then integrated into DefinitelyTyped? Last time I provided a type definition I did it on a fork of DefinitelyTyped and submitted a PR. Your project seems to be somewhat separate - so how do changes get back from here to TSD?

**Comment by goenning** [comment]
When `typings` got released, [people](https://github.com/typed-typings/) started to migrate some definitions from DefinitelyTyped into specific repositories, like this one. But things are recently changing a lot on how to type a package. 

Now that Microsoft annouced [The Future of Declaration Files](https://blogs.msdn.microsoft.com/typescript/2016/06/15/the-future-of-declaration-files/), I suppose that we should get back to update DefinitelyTyped. 

Is that correct @blakeembrey?

**Comment by blakeembrey** [comment]
@goenning I think I wrote a decent answer before: https://github.com/typings/typings/issues/600#issuecomment-226325643. The gist is, you're welcome to try and merge it back but I'd personally wait until the infrastructure is available. I will personally make sure that anyone using Typings is not left hanging, but we're waiting on redirects in DefinitelyTyped to exist (I don't particularly want to maintain my 200 definitions within DefinitelyTyped - that's 90% of the reason I tried to fix the model with Typings). 

By the way, @goenning, if you want I can add you to @typed-typings (if you would like).

**Comment by goenning** [comment]
Yes, sure @blakeembrey. I'll then move my repositories into it.

@0815fox When I get access to @typed-typings, I'll create a repository for pg-pool and we can keep the code there.

**Comment by blakeembrey** [comment]
Done 👍 You're welcome to create any repos you want there.

**Comment by goenning** [comment]
@0815fox https://github.com/typed-typings/npm-pg-pool is available.
after we get it into a releasable state we can add it to https://github.com/typings/registry

**Comment by 0815fox** [comment]
Okay, I am on it.

**Comment by 0815fox** [comment]
Okay, I started to create the typing (https://github.com/0815fox/npm-pg-pool), however, questions arise:
- pg-pool and pg seem to have circular dependency, as pg-pool uses the same interfaces (e.g. it returns an instance of Client). Should those be redefined in pg-pool or only defined in pg-pool?
- the module pg-pool exports a constructor function only. Am I right, that `export default class` is the correct way to describe that? Can/Should I still export further interfaces?
- in tests/pool.ts I started to define tests, however, I need the interfaces (which I originally did not export) in those tests. Shall I redefine them, declare vars which comply with them or export them in the module as well?
- In typed-pg I would add `export const Pool: new (config: Config) => PgPool;`. However, there seem to be further changes. So the question is: do you have an overview of what needs to be changed? Probably you have a better overview, as I just started to use node-postgres-module myself.

I am still a beginner, switched over from embedded development.
=> Sorry for the (maybe stupid) questions and thank you for your feedback.

**Comment by goenning** [comment]
`pg-pool and pg seem to have circular dependency, as pg-pool uses the same interfaces (e.g. it returns an instance of Client). Should those be redefined in pg-pool or only defined in pg-pool?`

Yeah, that sucks. But as TypeScript uses duck typing, I think we could duplicate the config parameters on each type definition for now.

`the module pg-pool exports a constructor function only. Am I right, that export default class is the correct way to describe that? Can/Should I still export further interfaces?`

Export only the Pool class, that'd be enought.

`in tests/pool.ts I started to define tests, however, I need the interfaces (which I originally did not export) in those tests. Shall I redefine them, declare vars which comply with them or export them in the module as well?`

Have you tried `import { default as Pool, Config } from "../index.d"`?

You sure you need to export them? look at typed-pg, I'm not exporting interfaces and I can actually usem them on my tests.

`In typed-pg I would add export const Pool: new (config: Config) => PgPool;. However, there seem to be further changes. So the question is: do you have an overview of what needs to be changed? Probably you have a better overview, as I just started to use node-postgres-module myself.`

I didn't look into the details of the changes, but  I saw that functions like connect and end has been moved to pool. Once we finish the pg-pool type we can add it as a dependency to pg and look at everything that changed here. One step at a time :+1:

**Comment by 0815fox** [comment]
Okay, thank you for the hints. Didn't notice, that you were importing things which you did not export into your tests - I did not even expect that to work...

I will continue work today in the evening or tomorrow.

Greetings from Switzerland.

**Comment by blakeembrey** [comment]
Just a note (since I'm still subscribed), but remember that the typings aren't `export default` - they should be `export =` to mirror the original source (https://github.com/brianc/node-pg-pool/blob/master/index.js#L6). CommonJS `module.exports =` has no mirror in ES6 module output, so you need to use `export =` to properly describe the typings.

**Comment by 0815fox** [comment]
Thank you @blakeembrey for the valuable input. Of course, you're right, - one also has to import the module itself using sth. like `import pgPool = require('node-pg-pool');`.

**Comment by 0815fox** [comment]
Ok, I just adapted it from `export default` to `export =`. However, then `import { default as Pool, Config } from "../index.d"` wont work anymore in the tests. If you agree, I will split up the interface definitions into a separate file and import them from index and tests.

What is the correct way for a module, which uses `export =` to export the interfaces as well?

**Comment by goenning** [comment]
That seems like a good option, I was looking for any example in existing repositiroes and didn't find any until now. 

But  why do we need the interfaces on our test? We'are only going to test Pool constructor and it's methods, right? So we would only to import it. Is that correct?

**Comment by blakeembrey** [comment]
Please don't split apart the files. Use an export with a namespace instead. For example:

``` ts
declare class Pool {}

declare namespace Pool {
  export interface Options {}
}

export = Pool
```

**Comment by 0815fox** [comment]
Okay, blake, I will do that.
The interfaces are needed outside as I am e.g. going to pass the client which is returned by the promise, to other methods. Those methods then need the Interface for declaring their arguments.
I thought checking method invocation in those tests would be easier when we can declare variables of the interface types. As we have to redeclare Interfaces like client, etc, we should probably also retest then, or am I wrong?

**Comment by blakeembrey** [comment]
No, seems good to me - making sure people can use the interfaces is definitely a nice to have. It's frustrating when I can't reuse a definition. In the example above, you can reuse easily enough by doing `import Pool = require('pg-pool'); const opts: Pool.Options = {}`.

**Comment by 0815fox** [comment]
Okay I close this one here, please continue discussion in pull request https://github.com/typed-typings/npm-pg-pool/pull/1
